libxrdp/libxrdp.c

```diff
@@ -540,14 +596,58 @@ libxrdp_send_bitmap(struct xrdp_session *session, int width, int height,
                 out_uint16_le(s, lines_sending);
                 out_uint16_le(s, bpp); /* bpp */
                 out_uint16_le(s, 0); /* compress */
-                out_uint16_le(s, (line_size + e * Bpp) * lines_sending); /* bufsize */
+                out_uint16_le(s, line_pad_bytes * lines_sending); /* bufsize */
                 q = p;
 
-                for (j = 0; j < lines_sending; j++)
+                switch (bpp)
                 {
-                    q = q - line_size;
-                    out_uint8a(s, q, line_size); /* B_ENDIAN doesn't work here, todo */
-                    out_uint8s(s, e * Bpp);
+                    case 8:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - line_bytes;
+                            out_uint8a(s, q, line_bytes);
+                            out_uint8s(s, e);
+                        }
+                        break;
+                    case 15:
+                    case 16:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((tui16*)(q + k * 2));
+                                out_uint16_le(s, pixel);
+                            }
+                            out_uint8s(s, e * 2);
+                        }
+                        break;
+                    case 24:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((tui32*)(q + k * 4));
+                                out_uint8(s, pixel);
+                                out_uint8(s, pixel >> 8);
+                                out_uint8(s, pixel >> 16);
+                            }
+                            out_uint8s(s, e * 3);
+                        }
+                        break;
+                    case 32:
+                        for (j = 0; j < lines_sending; j++)
+                        {
+                            q = q - server_line_bytes;
+                            for (k = 0; k < width; k++)
+                            {
+                                pixel = *((int*)(q + k * 4));
```

Comment:
Spacing:
`pixel = *((int *) (q + k * 4));`

I'm not a fan of `tui` types (we can change to `uint32_t` later), but let's use either `int` or `tui32` consistently in this function